### PR TITLE
How to document cnd's HTTP API

### LINKED ISSUE #1122: How to document the cnd HTTP API?
Answer the following questions:

- Which API-contracts approach should we use? (Originally taken from https://github.com/comit-network/comit-rs/issues/1221)
- How would a developer validate its API usage against these contracts? (Originally taken from https://github.com/comit-network/comit-rs/issues/1221)
- How to ensure the documentation and the implementation is not out of sync? Meaning examples actually work against the API (could be checked during e2e tests)
- How to generate the documentation (HTML files)?
- What does the documentation include?
- How to host documentation for current `master` HEAD?
- How to document extension link relation types (https://github.com/comit-network/comit-rs/issues/843)?

### PR DESCRIPTION
**Pull request by luckysori**
Fixes comit-network/comit-rs#1122.

I wrote this in emacs' org-mode and then automatically converted to `.adoc` so please also let me know if there's anything weird in terms of formatting.

### CODE REVIEW

File: 0015-how-to-document-cnd-http-api.adoc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+NOTE: Author: {authors} +
+Date: {revdate} +
+Tracking issue: https://github.com/comit-network/comit-rs/issues/1122[#1122]
+
+== Context ==
+
+We have no documentation for cnd's HTTP API. Having it is a hard requirement if we want to cater to application developers. It should also be helpful for the team and for newcomers.
+
+== Research ==
+We have no documentation for cnd's HTTP API. Having it is a hard requirement if we want to cater to application developers. It should also be helpful for the team and for newcomers.
```
Comment (bonomat) [inline 0015-how-to-document-cnd-http-api.adoc]: Duplication from above :)

File: 0015-how-to-document-cnd-http-api.adoc [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+=== What does the documentation include? ===
+
+==== HTTP API Specification. ====
+Models services.
+
+==== JSON Schema. ====
+Models data.
+
+=== How to host documentation for current `master` HEAD ===
+If we go for the recommended option of using custom-id:redoc[ReDoc], it's https://www.npmjs.com/package/redoc#tldr[very easy]:
```
Comment (da-kami) [inline 0015-how-to-document-cnd-http-api.adoc]: `custom-id:redoc[ReDoc]` does not render correctly

**Review by da-kami — APPROVED** [review]
Nice work, don't have anything to add :)

File: 0015-how-to-document-cnd-http-api.adoc (hunk 2 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: Thanks for pointing that out! Those are supposed to be internal links. I'll have to convert those manually to AsciiDoc.

File: 0015-how-to-document-cnd-http-api.adoc [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+:toc:
+:revdate: 2019-08-29
+
+NOTE: Author: {authors} +
+Date: {revdate} +
+Tracking issue: https://github.com/comit-network/comit-rs/issues/1122[#1122]
+
+== Context ==
+
+We have no documentation for cnd's HTTP API. Having it is a hard requirement if we want to cater to application developers. It should also be helpful for the team and for newcomers.
```
Comment (tcharding) [inline 0015-how-to-document-cnd-http-api.adoc]: One line per sentence :)

File: 0015-how-to-document-cnd-http-api.adoc [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+== Research ==
```
Comment (tcharding) [inline 0015-how-to-document-cnd-http-api.adoc]: nit: white space is non-uniform, sometimes a newline after a heading sometimes not.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 1 above)
Comment (tcharding) [inline 0015-how-to-document-cnd-http-api.adoc]: Duplicate lines.

File: 0015-how-to-document-cnd-http-api.adoc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+==== JSON Schema. ====
+Models data.
+
+=== How to host documentation for current `master` HEAD ===
+If we go for the recommended option of using custom-id:redoc[ReDoc], it's https://www.npmjs.com/package/redoc#tldr[very easy]:
+
+. Include API specification in repository with cnd.
+. Host a website on GitHub Pages (for example).
+. Use a HTML tag that links to the API specification (there's even a https://www.npmjs.com/package/redoc#usage-as-a-react-component[React component]).
```
Comment (tcharding) [inline 0015-how-to-document-cnd-http-api.adoc]: Should these dots be stars? `s/./*/`

**Review by tcharding — APPROVED** [review]
Approve with these few minor suggestions. Nice work.

File: 0015-how-to-document-cnd-http-api.adoc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+===== https://dredd.org/en/latest/index.html[Dredd] =====
+* Language-agnostic.
+* Works with both OpenAPI and API Blueprint.
+* Very popular (over 3k stars on GitHub).
+* Command line tool.
+* Takes API specification and makes requests to server based on that, reporting whether the responses match the documentation.
+* Example:
+[source,sh]
+----
+dredd cnd-http-api-description.yml http://localhost:<port-where-cnd-is-hosted>
```
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: So the API documentation would be enough for `dredd` to generate random-yet-valid queries. That's cool.

File: 0015-how-to-document-cnd-http-api.adoc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    expect(res).to.satisfyApiSpec;
+  });
+});
+----
+
+==== Recommendation ====
+Dredd, because it's well documented and has all the features we need. It looks like the main effort will be https://dredd.org/en/latest/how-it-works.html#making-your-api-description-ready-for-testing[preparing the specification] for testing.
+
+==== JSON Schema integration ====
+We currently use JSON Schema to validate the shape of the body of the response to `GET /
```
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: Wouldn't you expect to replace the JSON schema with our Open API doc?

File: 0015-how-to-document-cnd-http-api.adoc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+Models services.
+
+==== JSON Schema. ====
+Models data.
+
+=== How to host documentation for current `master` HEAD ===
+If we go for the recommended option of using custom-id:redoc[ReDoc], it's https://www.npmjs.com/package/redoc#tldr[very easy]:
+
+. Include API specification in repository with cnd.
+. Host a website on GitHub Pages (for example).
```
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: Would that allow developers to access the API specification of various versions or just master?

File: 0015-how-to-document-cnd-http-api.adoc (hunk 2 above)
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: what is this `custom-id:redoc[ReDoc]` ?

File: 0015-how-to-document-cnd-http-api.adoc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+. Use a HTML tag that links to the API specification (there's even a https://www.npmjs.com/package/redoc#usage-as-a-react-component[React component]).
+
+=== How to document extension link relation types? ===
+I would argue that this is covered in https://github.com/comit-network/comit-rs/issues/843[the original issue]. Just replace `human-protocol-spec` key with a link to a static HTML page hosted on GitHub Pages (for example) where the meaning of `human-protocol-spec` is described.
+
+=== Client-side validation ===
+
+==== Context ====
+* We already define custom-id:json-schema-integration[JSON Schemas] for some data objects returned from the API.
```
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: Is `custom-id:json-schema-integration[JSON Schemas]` supposed to be a link?

File: 0015-how-to-document-cnd-http-api.adoc [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+* Offer a copy of these contracts to clients in a programmatically accessible format.
+* Include link to in response header:
+[source,html]
+----
+Link: <http://example.com/schemas/swap.schema.json#>; rel=”describedby”
+----
+* Use in comit-i to prove that it works.
+
+==== References ====
+* https://apisyouwonthate.com/blog/the-many-amazing-uses-of-json-schema-client-side-validation[Blog on client-side validation based on JSON Schema].
```
Comment (D4nte) [inline 0015-how-to-document-cnd-http-api.adoc]: Could you descrive a bit more OpenAPI vs JSON schema? I am not clear on the concepts.

**Review by D4nte — APPROVED** [review]
Awesome job, few questions please and maybe some link issues?

File: 0015-how-to-document-cnd-http-api.adoc (hunk 3 above)
Comment (bonomat) [inline 0015-how-to-document-cnd-http-api.adoc]:
```suggestion
We have no documentation for cnd's HTTP API. Having it is a hard requirement if we want to cater to application developers. 
It should also be helpful for the team and for newcomers.
```

**Review by bonomat — APPROVED** [review]
Good job. 

Can you summarize all _sub recommendations_ in one **Recommendation** section please (with these subsections it's a bit hard to read)? 
So that if we refer back to this spike, we can see on one sight: 

* Use *OpenAPI* for the spec
* Use *Dredd* to ensure up2dateness
* ...

File: 0015-how-to-document-cnd-http-api.adoc (hunk 4 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: I think this doesn't matter unless you're looking at the raw file. Fixed it anyway :smile:

File: 0015-how-to-document-cnd-http-api.adoc (hunk 5 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: This is the way to render an [ordered list](https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#lists).

File: 0015-how-to-document-cnd-http-api.adoc (hunk 2 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: Broken link due to automatic conversion from org to asciidoc. Fixed now.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 9 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: Broken link due to automatic conversion from org to asciidoc. Fixed now.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 6 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: [Yeah](https://dredd.org/en/latest/how-it-works.html#how-it-works). But since you mention the word "random", it's important to point out that this is not fuzz-testing.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 7 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: My understanding is that JSON Schema is more expressive and can be used for things such as server-defined client-side validation. OpenAPI supports a subset of JSON Schema and in the [near future](https://github.com/OAI/OpenAPI-Specification/pull/1977) if will be fully supported.

In any case, using the `$ref` keyword we can reference `.schema.json` files in OpenAPI files. Until OpenAPI supports JSON Schema in full there might be some pain points, but there are ways to [work around](https://blog.apisyouwonthate.com/openapi-and-json-schema-divergence-part-2-52e282e06a05) them.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 8 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: Talking to @thomaseizinger, we see the publicly hosted API specification only supporting master.

For someone developing against the API locally, we recommend adding an endpoint which returns the specification to the HTTP API itself.

File: 0015-how-to-document-cnd-http-api.adoc (hunk 10 above)
Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: Hopefully clearer now.

**Review by thomaseizinger — APPROVED** [review]
Awesome, good work! :)

The resource (`/http-spec`) on that hosts the documentation of `cnd` is probably debatable, but that is a minor nit :grimacing:

File: 0015-how-to-document-cnd-http-api.adoc (hunk 10 above)
Comment (D4nte) [reply 0015-how-to-document-cnd-http-api.adoc]: What commit made it clearer?

Comment (luckysori) [reply 0015-how-to-document-cnd-http-api.adoc]: [This one](https://github.com/comit-network/spikes/pull/38/commits/e23ed624e1ce84896b1a23e3e0cf3faa61ab9122).
